coap-gateway/service/signIn.go

```diff
@@ -168,6 +170,62 @@ func setNewDeviceSubscriber(ctx context.Context, client *Client, deviceID string
 	return nil
 }
 
+func (client *Client) updateBySignInData(ctx context.Context, deviceId, userId, accessToken string, validUntil time.Time) error {
+	authCtx := authorizationContext{
+		DeviceID:    deviceId,
+		UserID:      userId,
+		AccessToken: accessToken,
+		Expire:      validUntil,
+	}
+	oldAuthCtx := client.SetAuthorizationContext(&authCtx)
+
+	if err := client.server.devicesStatusUpdater.Add(client); err != nil {
+		return fmt.Errorf("cannot update cloud device status: %w", err)
+	}
+
+	newDevice := false
+	switch {
+	case oldAuthCtx.GetDeviceID() == "":
+		newDevice = true
+	case oldAuthCtx.GetDeviceID() != deviceId || oldAuthCtx.GetUserID() != userId:
+		newDevice = true
+		client.cancelResourceSubscriptions(true)
+		if err := client.closeDeviceSubscriber(); err != nil {
+			log.Errorf("failed to close device %v connection: %w", oldAuthCtx.GetDeviceID(), err)
+		}
+		client.cleanObservedResources()
+		client.unsubscribeFromDeviceEvents()
+	}
+
+	if newDevice {
+		if err := client.loadShadowSynchronization(ctx, deviceId); err != nil {
+			return fmt.Errorf("cannot load shadow synchronization for device %v: %w", deviceId, err)
+		}
+
+		if err := setNewDeviceSubscriber(ctx, client, deviceId); err != nil {
+			return fmt.Errorf("cannot set device subscriber: %w", err)
+		}
+
+		if err := client.subscribeToDeviceEvents(userId, func(e *events.Event) {
+			evt := e.GetDevicesUnregistered()
+			if evt == nil {
+				return
+			}
+			if evt.Owner != userId {
+				return
+			}
+			if strings.Contains(evt.DeviceIds, deviceId) {
+				if err := client.Close(); err != nil {
+					log.Errorf("sign in error: cannot close client: %w", err)
+				}
+			}
+		}); err != nil {
+			return fmt.Errorf("cannot subscribe to device events: %w", err)
+		}
```

Comment:
you need to recheck if `OwnsDevice` contains the device after creation a subscription - it can be removed between first get `OwnsDevice` and  `Subscribe`.